File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.js
const input = require("fs")
  .readFileSync("/dev/stdin", "utf8")
  .trim()
  .split("\n")
  .map((line) => line.replace(/\r/, ""));

// -1, +1, *2로 이동이 가능하다
// 0<x<100000
// 1초씩 걸리고 가장빠른 초
// [위치, 시간]으로 bfs로 탐색하면서 set에 넣고 중복 체크한다
const solution = (input) => {
  const [start, end] = input[0].split(" ").map((v) => Number(v));
  const set = new Set();
  set.add(start);

  const queue = [[start, 0]];

  while (queue.length) {
    const [pos, time] = queue.shift();

    if (pos === end) {
      console.log(time);
      break;
    }

    if (pos - 1 >= 0 && !set.has(pos - 1)) {
      queue.push([pos - 1, time + 1]);
      set.add(pos - 1);
    }

    if (pos + 1 <= 100000 && !set.has(pos + 1)) {
      queue.push([pos + 1, time + 1]);
      set.add(pos + 1);
    }

    if (pos * 2 <= 100000 && !set.has(pos * 2)) {
      queue.push([pos * 2, time + 1]);
      set.add(pos * 2);
    }
  }
};

solution(input);

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.py
from collections import deque

# 첫번째줄에 시작점 도착점이 주어진다 1초후에 -+1, 두배 위치로 이동 가능하다
# bfs로 방문한 지점 마킹하면서 100000보다 작으면 된다

start, end = list(map(int, input().split()))

queue = deque()
queue.append([start, 0])
visited = [False]*(100001)

while len(queue):
  pos, cnt = queue.popleft()
  if pos == end:
    print(cnt)
    break
  
  if pos*2 <= 100000 and not visited[pos*2]:
    queue.append([pos*2, cnt +1])
    visited[pos*2] = True

  if pos - 1 >= 0 and not visited[pos-1]:
    queue.append([pos-1, cnt + 1])
    visited[pos-1] = True

  if pos + 1<= 100000 and not visited[pos + 1]:
    queue.append([pos+1, cnt + 1])
    visited[pos+1] = True
